File: DSA09029.py
from collections import defaultdict


def function():
    v, e = map(int, input().split())
    ls = []
    while len(ls) < 2 * e:
        ls += list(map(int, input().split()))

    adj = defaultdict(list)
    for i in range(e):
        adj[ls[2*i]] .append(ls[2*i + 1])
        adj[ls[2*i + 1]] .append(ls[2*i])

    visited = defaultdict(bool)

    def hamilton(l, i):        
        if l == v:
            return 1

        else:
            for x in adj[i]:
                if not visited[x]:
                    visited[i] = True
                    if hamilton(l + 1, x):
                        return 1
                    visited[i] = False
        return 0

    visited[1] = True
    if hamilton(1, 1):
        return 1
    return 0

if __name__ == "__main__":
    for i in range(int(input())):
        print(function())

File: DSA09029.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;
void hamilton(int l, int i, int v, unordered_map<int, vector<int>>& adj, vector<bool>& visited, bool& found) {
    if (l == v) {
        found = true;
    } else {
        for (int x : adj[i]) {
            if (!visited[x]) {
                visited[i] = true;
                hamilton(l + 1, x, v, adj, visited, found);
                visited[i] = false;
            }
        }
    }
}

int main() {
    int t;
    cin >> t;
    while (t--) {
        int v, e;
        cin >> v >> e;
        unordered_map<int, vector<int>> adj;
        vector<bool> visited(v + 1, false);
        int a, b;
        for (int i = 0; i < e; ++i) {
            cin >> a >> b;
            adj[a].push_back(b);
            adj[b].push_back(a);
        }
        bool found = false;
        for (int i = 1; i <= v + 1; i++) {
            hamilton(1, i, v, adj, visited, found);
            if (found) break;
        }
        cout << found << endl;
    }
    return 0;
}
